src/poetry/core/factory.py

```diff
@@ -622,6 +638,35 @@ def validate(
 
         return result
 
+    @classmethod
+    def _validate_dependency_groups_includes(
+        cls, toml_data: dict[str, Any], result: dict[str, list[str]]
+    ) -> None:
+        """Ensure that dependency groups do not include themselves."""
+        config = toml_data.setdefault("tool", {}).setdefault("poetry", {})
+
+        group_includes = {}
+        for group_name, group_config in config.get("group", {}).items():
+            if include_groups := group_config.get("include-groups", []):
+                group_includes[canonicalize_name(group_name)] = [
+                    canonicalize_name(name) for name in include_groups
+                ]
+
+        for group_name in group_includes:
+            visited = set()
+            stack = [group_name]
+            while stack:
+                group = stack.pop()
+                visited.add(group)
+                for include in group_includes.get(group, []):
+                    if include in visited:
+                        result["errors"].append(
```

Comment:
I think this algorithm will provide a false error if presented a specific diamond pattern:

* a
  * b1
     * shared 
  * b2
    * c
       * shared

It'll visit `a`, `b1`, `b2`, `shared`, and then it will fail on `c`.

---

In testing, this also fails with:


* a
  * b1
     * shared 
  * b2
     * shared

---

I think the simplest algorithm is to keep a list of ancestors for every group you visit, verifying there are no duplicates.

Reply:
Fixed: https://github.com/finswimmer/core/pull/2